pkg/credentialprovider/apis/dockerauth/types.go

```diff
@@ -0,0 +1,100 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockerauth
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DockerAuthPluginConfig is the configuration containing information about
+// each docker auth plugin. This configuration type is consumed by the kubelet.
+type DockerAuthPluginConfig struct {
+	metav1.TypeMeta
+
+	// Plugins is a list of docker auth plugins that will be enabled by the kubelet
+	Plugins []DockerAuthPlugin
+}
+
+// DockerAuthPlugin is used by the kubelet container runtime to match the
+// image property string (from the container spec) with exec-based credential provider
+// plugins that provide container registry credentials.
+type DockerAuthPlugin struct {
+	// name is the name of the credential plugin.  It must match the name of the
+	// executable as seen by the kubelet.
+	Name string
+
+	// imageMatchers is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// DockerAuthPlugin will be used by kubelet to provide credentials
+	// for the image pull.
+
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each ImageMatcher string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
+	ImageMatchers []string
+
+	// ExtraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg
+}
+
+// PluginArg is used for passing extra arguments to the exec credential plugin binary.
+type PluginArg struct {
+	Name  string
+	Value string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DockerAuthPluginRequest is passed to the plugin via stdin, and includes the image that will be pulled by kubelet.
```

Comment:
If this is passed to the plugin, will the plugin want to be able to import this struct? (If so we probably want it in staging not pkg)

Reply:
Yeah +1, plugin will definitely want to import this without redefining it.

Reply:
+1 to this, there's a section in the KEP about moving this to staging https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md#moving-credential-providers-to-staging 

I'm hoping this can be done once we get an alpha implementation in first.